File: tests/workflow/consumer/rust/src/test/test_request_groupb_groupc_structa.rs
use super::{controller, samples, stat, ClientError, Consumer, StatEvent};
use tokio::sync::mpsc::UnboundedSender;

pub async fn execute(
    consumer: &mut Consumer<ClientError>,
    tx_stat: &UnboundedSender<StatEvent>,
) -> Result<(), String> {
    let mut struct_a = false;
    let mut struct_b = false;
    while !struct_a || !struct_b {
        let response = consumer
            .groupb_groupc_structa(samples::group_b::group_c::struct_a::get())
            .await
            .map_err(|e| e.to_string())?;
        tx_stat
            .send(StatEvent::Inc(stat::Alias::TestRequestGroupBGroupCStructA))
            .map_err(|e| e.to_string())?;
        match response {
            controller::GroupBGroupCStructAResponse::Response(res) => {
                if !samples::group_b::group_c::struct_b::equal(res.clone()) {
                    return Err(format!("Invalid data received: {:?}", res));
                }
                struct_a = true;
            }
            controller::GroupBGroupCStructAResponse::Err(res) => {
                if !samples::group_a::struct_b::equal(res.clone()) {
                    return Err(format!("Invalid data received: {:?}", res));
                }
                struct_b = true;
            }
        };
    }
    Ok(())
}

File: tests/workflow/consumer/typescript/src/test/test_request_groupb_groupc_structa.ts
// tslint:disable: variable-name
import { Protocol, GroupBGroupCStructA } from "../consumer/index";
import { Consumer } from "../consumer";

export function test(consumer: Consumer): Promise<void> {
	return new Promise((resolve, reject) => {
		let case_GroupAStructB = false;
		let case_GroupBGroupCStructB = false;
		function send(): Promise<void> {
			return new Promise((res, rej) => {
				const request = new GroupBGroupCStructA(
					Protocol.GroupB.GroupC.StructA.defaults(),
					consumer
				);
				request.send().then((response) => {
					if (response instanceof Protocol.GroupA.StructB) {
						if (!case_GroupAStructB) {
							case_GroupAStructB = true;
							res(undefined);
						} else {
							rej(
								new Error(
									`Multiple times get: Protocol.GroupA.StructB`
								)
							);
						}
					} else if (
						response instanceof Protocol.GroupB.GroupC.StructB
					) {
						if (!case_GroupBGroupCStructB) {
							case_GroupBGroupCStructB = true;
							res(undefined);
						} else {
							rej(
								new Error(
									`Multiple times get: Protocol.GroupB.GroupC.StructB`
								)
							);
						}
					}
				});
			});
		}
		function trigger() {
			send()
				.then(() => {
					if (case_GroupBGroupCStructB && case_GroupAStructB) {
						resolve(undefined);
					} else {
						trigger();
					}
				})
				.catch(reject);
		}
		trigger();
	});
}
